### PR TITLE
[relations]: add reordering keyboard navigation

### PR DESCRIPTION
**Pull request by joshuaellis**
### What does it do?

* Adds keyboard-navigation for re-ordering relations
* Adds live region for announcing the actions during re-ordering

### Context
This follows the aria-spec to implement keyboard navigation with re-ordering that I had to source for numerous places, namely Atlassian's approach (which makes perfect sense imo). It's designed this way _because_ we have clickable areas inside our relations i.e. the link to the entity, therefore you need a button to attach too. It's also added this way in the `adobe spectrum` project.

### Why is it needed?

* Because some user's might not be able to use their mouse.

### How to test it?

* There are automated tests

### Related issue(s)/PR(s)

* should be merged after #14800
* resolves CONTENT-656

### CODE REVIEW

**Review by jhoward1994 — CHANGES_REQUESTED** [review]
I'm having an issue where I can't click on drag handle to move a relation around, only the main content of the relation item.

It seems to be because the click event of `IconButton` is fired on mouse up rather than mouse down.

**Comment by joshuaellis** [comment]
> I'm having an issue where I can't click on drag handle to move a relation around, only the main content of the relation item.
> 
> It seems to be because the click event of `IconButton` is fired on mouse up rather than mouse down.

Hey @jhoward1994 what browser are you using? 🤔 I don't see this issue on my end rn but with more info i'll look into it 🕵🏼

**Comment by jhoward1994** [comment]
> > I'm having an issue where I can't click on drag handle to move a relation around, only the main content of the relation item.
> > It seems to be because the click event of `IconButton` is fired on mouse up rather than mouse down.
> 
> Hey @jhoward1994 what browser are you using? 🤔 I don't see this issue on my end rn but with more info i'll look into it 🕵🏼

Aha! I was testing with Firefox, it seems to be working for me in Brave/Chrome too

**Review by jhoward1994 — APPROVED** [review]
Looks good, working well for me now on all browsers

File: packages/core/admin/admin/src/content-manager/components/RelationInput/RelationInput.js [hunk 1]
```diff
@@ -312,14 +371,25 @@ const RelationInput = ({
         }
       >
         <RelationList overflow={overflow}>
+          <VisuallyHidden id={`${name}-item-instructions`}>
+            {formatMessage({
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/RelationInput/RelationInput.js:375]: Can we please pass that through as a prop? If we decide to move the component into the DS we shouldn't have a dependency on react-intl.

**Review by gu-stav — COMMENTED** [review]
The code looks good to me. Only two minor comments. During the test I found the following things a bit weird (they are not related to the PR but I wanted to mention them anyways):

- Icon outline overflows the label

Screenshot 2022-11-11 at 11 54 30

- If there is only one item the handler is not visible: I understand where this is coming from, but I don't like the visual shift tbh. What do you think about checking with Maeva to see if we should disable it instead?

Screenshot 2022-11-11 at 11 56 02

- the state is already updated on `drag` which causes the `save` button to flicker (from disabled to enabled) even though I have not dropped the item yet. Do you think it is worth doing something about that?

**Comment by joshuaellis** [comment]
> the state is already updated on drag which causes the save button to flicker (from disabled to enabled) even though I have not dropped the item yet. Do you think it is worth doing something about that?

Whilst I agree this is a bit annoying, this would require more of a refactor i.e. we have to clone (as opposed to use) `modifiedData` state into `RelationInputDataManager`component and use that and then at point X replace that in the store.
IMO this is a problem with our implementation of the `save` button, a better way to do it in the future could be "canSave" in the store so we can have better control over when it changes as opposed to see if there are changes in `modifiedData`.

> If there is only one item the handler is not visible: I understand where this is coming from, but I don't like the visual shift tbh. What do you think about checking with Maeva to see if we should disable it instead?

cc @maevalienard for some design guidance.

> Icon outline overflows the label

This is because the `IconButton` has it's own padding which the design does not adhere too. cc @maevalienard for more design guidance

File: packages/core/admin/admin/src/content-manager/components/RelationInput/RelationInput.js [hunk 2]
```diff
@@ -392,15 +412,21 @@ RelationInput.propTypes = {
   error: PropTypes.string,
   description: PropTypes.string,
   disabled: PropTypes.bool,
+  iconButtonAriaLabel: PropTypes.string,
   id: PropTypes.string.isRequired,
   label: PropTypes.string.isRequired,
   labelAction: PropTypes.element,
   labelLoadMore: PropTypes.string,
   labelDisconnectRelation: PropTypes.string.isRequired,
+  listAriaDescription: PropTypes.string,
```
[the remark this answers is not in the document]
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/components/RelationInput/RelationInput.js:421]: > pass in the translated string

we do already I believe. But yes, we can make them required & remove their defaults.

**Comment by maevalienard** [comment]
> If there is only one item the handler is not visible: I understand where this is coming from, but I don't like the visual shift tbh. What do you think about checking with Maeva to see if we should disable it instead?

Interesting. Here are my thoughts:
- For one-way relations: no icon displayed at all.
- For many-ways relations: icon always displayed. Even if you only have selected one relation. And no disable at all, because that's still fun to play around, we could benefit from the discoverability :) 

> Icon outline overflows the label
I assume the drag&drop icon has been added via an iconButton so there are no possibilities to actually reduce the hit zone of this component for this usecase? I think we could consider having a slightly more padding-left on the relation label to avoid that. What do you think?

File: packages/core/admin/admin/src/content-manager/components/RelationInput/components/RelationItem.js [hunk 3]
```diff
@@ -10,26 +10,37 @@ import { IconButton } from '@strapi/design-system/IconButton';
...
 const ChildrenWrapper = styled(Flex)`
   width: 100%;
   /* Used to prevent endAction to be pushed out of container */
   min-width: 0;
 `;
 
+const IconButtonWrapper = styled.div`
+  margin-right: ${(props) => props.theme.spaces[1]};
+`;
+
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/RelationInput/components/RelationItem.js:24]: What do you think about using `Stack spacing={1}` as a wrapper below instead?

File: packages/core/admin/admin/src/content-manager/components/RelationInputDataManager/RelationInputDataManager.js [hunk 4]
```diff
@@ -346,6 +346,7 @@ export const RelationInputDataManager = ({
         'isSuccess'
       )}
       required={required}
+      toOneRelation={toOneRelation}
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/RelationInputDataManager/RelationInputDataManager.js:349]: What do you think about passing `canReorder` down instead? It makes the implementation a bit more independent ...

**Review by gu-stav — CHANGES_REQUESTED** [review]
It works really well for me now! Thanks.

I've noticed a visual glitch though, if the content of the main field is too long:

Screenshot 2022-11-14 at 16 14 46

Also: the `Status` is no longer aligned at the end.

**Comment by joshuaellis** [comment]
Thanks, I assume this is to do with wrapping the icon and description in Stack. 

I'll take a look to see if it's easily solvable, otherwise I'll revert back to the previous solution of using a wrapper.

**Review by ronronscelestes — CHANGES_REQUESTED** [review]
great work 🔥
still seeing an issue with alignment of the status badge on my side 😬 status should be stacked next to the cross icon: 
image

no more overflow issue though 💃
image

**Review by ronronscelestes — APPROVED** [review]
LGTM ✨🙏

**Pull request merged**